src/aws-manager.js

```diff
@@ -139,10 +125,172 @@ class AwsManager {
     this.reportAmiUsage = series.amiUsage.reporter(influx);
   }
 
+  async init(){
+    try {
+      let rawJson = await this.spotRequestContainer.read('internal-provisioner-data');
+      let data = JSON.parse(rawJson);
+      // This is mainly for debugging
+      fs.writeFileSync('internal-data.json', data);
+      this.__spotRequestIdCache = data.managedSpotRequests || []; 
+      this.__internalState = data.internallyTrackedSpotRequests || [];
+    } catch (err) {
+      if (err.code !== 'BlobNotFound') {
+        throw err;
+      }
+      this.__spotRequestIdCache = [];
+      this.__internalState = [];
+    }
+  }
+
+  async saveAwsManagerInternalState() {
+    // We use this for both of the current internal caches... let's just use
+    // the same code for both
+    function expire(item) {
+      let diff = Date.now() - item.created;
+      // If we've known about this for more than 5 days, evict it, otherwise
+      // keep it.  5 days is a good amount since we force kill things even if
+      // they're running after 4 days.  This lets us have a full extra day
+      if (diff > 1000 * 60 * 60 * 24 * 5) {
+        return false;
+      }
+      return true;
+    }
+
+    this.__spotRequestIdCache = this.__spotRequestIdCache.filter(expire);
+
+    let data = JSON.stringify({
+      managedSpotRequests: this.__spotRequestIdCache || [],
+      internallyTrackedSpotRequests: this.__internalState || [],
+    }, null, 2);
+
+    fs.writeFileSync('internal-data.json', data);
+    return this.spotRequestContainer.write('internal-provisioner-data', data);
+  }
+
+  /**
+   * Simplify how we access stored spot request ids
+   */
+  async workerTypeForResource(resource, region) {
+    assert(typeof resource === 'object');
+    assert(typeof region === 'string');
+
+    let srid = resource.SpotInstanceRequestId;
+    let instanceId = resource.InstanceId;
+
+    // First, we should check to see if we have the mapping of
+    // SpotInstanceRequestId in memory (which we persist to an azure blob
+    // storage blob).  If this item is in that cache, we can be sure that the
+    // provisioner both owns it because it created it, and also which worker
+    // type, since we add that information to the cache when we add the id.
+    let found = this.__spotRequestIdCache.filter(x => x.id === srid && x.region === region);
+    if (found.length === 1) {
+      return found[0].workerType;
+    } else if (found.length > 1) {
+      let firstX = found[0].workerType;
```

Comment:
Isn't it an invariant violation if there is more than one?

IMO we should just do: 
```js
let entry = this.__spotRequestIdCache.filter(x => x.id === srid && x.region === region).pop();
if (entry) {
  return entry.workerType;
}
```

And when we load `__spotRequestIdCache` as well as after each iteration we do some sanity checks. To ensure the invariants are enforced.
Example:
```js
let uniqueEntries = _.uniqBy(this.__spotRequestIdCache, entry => entry.region + '/' +entry.id).length;
assert(uniqueEntries === this.__spotRequestIdCache.length, 'Duplicate entries in __spotRequestIdCache');
```

This is much less code and enforces our invariants much better... If we include the assert above after loading state, before saving state (so we never save bad state), and after modifying the state in an iteration.. Then we're in a really good place.

We could probably just introduce a simple `this._validateState()` method that validates all the internal state.. Sends the an error to sentry if there is a violation... Then we can also use `this._validateState()` to drop current state if there is an invariant violation. We can do this simple thing after every iteration, after/before loading/saving state, etc... 

Having a few check points where we validate all invariants of the internal state is a lot simpler.